### PR TITLE
Lf 2688 cypress tests for tasks

### PR DESCRIPTION
**Pull request by MwayaB**
**Description**

This PR includes more robust testing of the tasks flow and adds a husky pre-push hook to run said test suite in the developers local environment

Jira link:

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Passes test case
- [ ] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

### CODE REVIEW

File: packages/webapp/cypress/e2e/tasksFlow.spec.js [hunk 1]
```diff
@@ -0,0 +1,750 @@
...
+      const fieldName = 'Test Field';
+      const workerName = 'John Worker';
+      const testCrop = 'New Crop';
+      const role = 'Manager';
+      const lang = 'English';
+      let taskType_id;
+
+      const date = new Date();
+      //Inputs for crop plan
+      const daysGermination = 10;
```
Comment (Duncan-Brain) [inline packages/webapp/cypress/e2e/tasksFlow.spec.js:43]: A lot of these constants can be and are already imported with cy.fixtures I believe

File: packages/webapp/cypress/e2e/tasksFlow.spec.js [hunk 2]
```diff
@@ -0,0 +1,750 @@
...
+      .and('not.be.disabled')
+      .click({ force: true });
+
+    cy.selectTaskLocation();
+    cy.get('[data-cy="addTask-cropsContinue"]').click();
+    cy.get('[data-cy="react-select"]').type(`Other{enter}`);
+    cy.get('[data-cy="irrigateTask-type"]').type(customTask);
+    cy.get('[type = "checkbox"]').eq(0).check({ force: true });
+    cy.get('[type = "checkbox"]').eq(1).check({ force: true });
+    //cy.get('[data-cy="irrigateTask-calculate"]').click({ force: true });
```
[the remark this answers is not in the document]
Comment (MwayaB) [reply packages/webapp/cypress/e2e/tasksFlow.spec.js:306]: @Duncan-Brain Understood

**Comment by Duncan-Brain** [comment]
@MwayaB Thank you for your continued work on cypress tests! This is a really tough problem so thanks for taking it on!

So this definitely looks like a step in the right direction. I did a test push and the cypress tests successfully ran and then pushed my test branch to Github.

Sorry if this is a lot comments but this is a decent sized PR with a lot of implications! My thoughts so far:

1. Husky pre-push
- I would suggest moving this to a different ticket and us all having a discussion about it if it is what you and Kevin think is best
- I did a push test and it took about 5 minutes to run  -- while I like the idea of getting feedback early in the development process I have doubts that this makes sense after every push to github. Especially when tasks are not being worked on for a ticket.
- I personally prefer the idea of building out an optional npm script

2. Naming conventions
- this is a small point but it might be good to choose a few naming conventions -- it looks like for data-cy what I am loosely seeing is <componentName>-<label/formName> . It might be good to commit to a specific way no matter how long it gets haha (I don't love what it means though heres a super long example.. waterUsageCalculatorModal-estimatedWaterUsageUnit)
- for fixtures files is there a structure you are following for creating these like tasksFlow, happyPathFlow, cropPlanFlow? Sometimes these constants files can get really big and I was wondering if it made sense to base the constants file names based on tables in the db.

3. Inline comments
- commented out code
- moving fixtures constants

4.  beforeEach
- just wondering if it is necessary to log out and log in before each task test? It may be valuable to only do this occasionally.

5. Jira ticket scope
- When I found the Jira ticket the scope says: Test every task type creation, abandon, and completion flows. But it looks like this PR tests creation?
- This might be a really big PR if you have to add abandon and completion flows too it may be helpful to chunk this up into smaller tickets?
- The Jira ticket didn't say much about the task tests itself and what needs to be tested is there a document somewhere?

6. General
- In my opinion big PRs make it hard for me to get fast and regular "wins" and make it easier for others test so it might be worth considering
- The PR description both here and on the Jira ticket could maybe use some love -- since we all need to come up to speed with cypress it could be useful to share how you tested it here and what the test requirements are for tasks probably on the Jira ticket

Sorry if this was a lot but it looks good and is coming a long nicely by the looks of it!

**Comment by MwayaB** [comment]
@Duncan-Brain 
1. Husky pre-push
- I agree that this does not necessarily fall within the scope of this ticket. But i would like to provide some context, I arrived at the idea of running e2e tests locally after having experienced a lot of flakiness running this test suite on the CI runner . Kevin and I have not discussed this approach, it is something i mentioned that i was going to attempt in standup on May 3rd and since i had never done this before decided this ticket was a good candidate to attempt a proof of concept.
- My thinking was exactly that, to give the developer early feedback instead of them waiting until they create a PR to the integration branch, the idea was to not only have the tasks user flow test but to progressively add test suites for other user flows in the app as and when those become available. There was also the point i made above about my experience where tests ran more consistently and faster in my local environment
- It is my understanding that we want automated tests to run as a default part of the development process and not an option, however the pre-push hook can still be omitted by using the `--no-verify` flag. An obvious issue here is that if said tests are to run locally we only have 2 options the pre-commit and pre-push hooks
2. Naming conventions
-  Agreed i have tried as much as possible to stick to this naming convention however there are cases where  i made the decision to use  pre-existing data-testid tags, which do not always follow this naming instead of adding new data-cy tags, there are also some cases where i used some data-cy tags that were added by the developer and these may not always follow the established convention.
- For fixtures, the actual files are named after the user flow without the 'flow' part so 'tasks' for the fixture file used in the tasksFlow suite. For constants since i am approaching these tests from the front end i have used the name of the input displayed on the frontend.
4. I have the log out/in in the before each hook to ensure that state is reset after each test
5. Jira ticket scope
-  Thanks, i was working towards the initial scope '70% test coverage of the webapp'.
- Agreed adding abandonment, completion and deletion? would make the PR significantly larger
- No, i am not aware of any document

**Comment by MwayaB** [comment]
@Duncan-Brain Well getting happy path to run at all in CI required quite a lot of refactoring and commenting of the intended test suite which significantly reduced its code coverage, after my experience trying to run the tasks suite in CI i believe something similar would be required.
1. Thats fair
2. Yes

**Pull request closed without merging**